enhancements/kube-apiserver/enforce-labels.md

```diff
@@ -0,0 +1,123 @@
+---
+title: enforce-labels
+authors:
+ - "@mmirecki"
+reviewers:
+ - "@fedepaol"
+approvers:
+ - TBD
+creation-date: 2020-08-07
+last-updated: 2020-08-07
+status: implementable
+---
+
+# Enforced Labels and Annotations
+
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [x] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement intends to provide a way to define, per project, labels and annotations that should be enforced in all resources of a given type created in the particular project (v.s. having those possibly  pre-defined in helm chart). The labels could be specified per project, resource type, resource name.
+
+## Motivation
+
+With the deployment of complex applications in multi-cluster and multi-tenant configurations, organizations use labels and annotations to represent complex organizational dependencies, a state in an approval or validation process, or other metadata the organization uses in internal processes.
+
+As telco platforms are still evolving, so are the labels requirements. Those are likely to increase in volume. Managing those through the helm chart is not practical for CNF vendors, as they would need to maintain a separate chart for each provider. Moreover, this would only apply to those vendors that deliver their CNF via helm charts.
+
+As an example, some providers expect CaaS specific labels in order to apply the service proxy configurations. These labels are used in the Service Provider’s environment to inject sidecars for service mesh configuration. Other possible uses can be logging or monitoring.
+
+### Goals
+
+- enforce a per project (namespace) labeling scheme
+- ability to configure a cluster-level labeling scheme to enforce on all or a subset of namespaces
+- specify the labeling scheme on resource type and resource level
+- the labeling should complement installation time (eg. Helm or Operator) labeling scheme
+- ensure the labeling scheme is maintained over resource lifecycle (e.g. pod restart)
+- allow customer organizations to update the labeling scheme at any time
+- relabel existing resources upon labeling scheme update
```

Comment:
including removal?

Reply:
Technically it is possible. Rego (in which the constraints are written) would allow this.
There are still issues like rule conflicts or rule convergence which need to be addressed. There is already some discussion ongoing on this on the proposal issue (https://github.com/open-policy-agent/gatekeeper/issues/809).
This feature has no clear ETA partly because there is no clear understanding on how to tackle this.